src/swift/Sources/TokenGeneration/NavigationTokens.swift

```diff
@@ -0,0 +1,149 @@
+        guard publicModifiers.contains(decl.accessLevelModifier ?? .internal) else {
+            return nil
+        }
+        let navItem = NavigationItem(text: decl.name.textDescription, navigationId: decl.name.textDescription, typeKind: .class)
+        decl.members.forEach { member in
+            switch member {
+            case let .declaration(decl):
+                if let item = navigationTokens(from: decl) {
+                    navItem.childItems.append(item)
+                }
+            case .compilerControl:
+                return
+            }
+        }
+        return navItem
+    }
+
+    private func navigationTokens(from decl: EnumDeclaration) -> NavigationItem? {
+        guard publicModifiers.contains(decl.accessLevelModifier ?? .internal) else {
+            return nil
+        }
+        let navItem = NavigationItem(text: decl.name.textDescription, navigationId: decl.name.textDescription, typeKind: .class)
+        decl.members.forEach { member in
+            switch member {
+            case let .declaration(decl):
+                if let item = navigationTokens(from: decl) {
+                    navItem.childItems.append(item)
+                }
+            default:
+                return
+            }
+        }
+        return navItem
+    }
+
+    private func navigationTokens(from decl: ProtocolDeclaration) -> NavigationItem? {
+        guard publicModifiers.contains(decl.accessLevelModifier ?? .internal) else {
+            return nil
+        }
+        let navItem = NavigationItem(text: decl.name.textDescription, navigationId: decl.name.textDescription, typeKind: .class)
+        return navItem
+    }
+
+    private func navigationTokens(from decl: ExtensionDeclaration) -> NavigationItem? {
+        guard publicModifiers.contains(decl.accessLevelModifier ?? .internal) else {
+            return nil
+        }
+        let navItem = NavigationItem(text: decl.type.textDescription, navigationId: decl.type.textDescription, typeKind: .class)
+        decl.members.forEach { member in
+            switch member {
+            case let .declaration(decl):
+                if let item = navigationTokens(from: decl) {
+                    navItem.childItems.append(item)
+                }
+            default:
+                return
+            }
+        }
+        return navItem
+    }
+
+    private func navigationTokens(from decl: Statement) -> NavigationItem? {
+        switch decl {
+        case let decl as ClassDeclaration:
+            return navigationTokens(from: decl)
+        case let decl as EnumDeclaration:
+            return navigationTokens(from: decl)
+        case let decl as ExtensionDeclaration:
+            return navigationTokens(from: decl)
+        case let decl as ProtocolDeclaration:
+            return navigationTokens(from: decl)
+        case let decl as StructDeclaration:
+            return navigationTokens(from: decl)
+        default:
+            return nil
+        }
```

Comment:
Might want to consider conforming certain declarations to a protocol to make the function less verbose. Not too much of the navigation is declaration-specific.

Reply:
That's true, but then you'd need extensions for each declaration type to be `NavigationTokenizable`. so you save lines here but then duplicate them elsewhere.

Reply:
The problem is that `Declaration` itself is an empty protocol and so you can't do anything with it, hence the need for these silly switch statements.